File: A/UVA/OrderingTasks.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> result;
vector<vector<int>> graph;
bool visited[110] ;

void toposort(int parent){
    visited[parent] =true;
    int sz = graph[parent].size();
    for(int i = 0;i < sz;i++){
        int child = graph[parent][i];
        if (!visited[child]){
            toposort(child);
        }
    }
    result.push_back(parent);
}

int main(){
    int n,m;
    cin >> n >> m ;
    graph.resize(n+1);
    for(int i = 0 ; i < m+1 ;i++){
        int s,e;
        cin >> s >> e;
        if (s == 0) break;
        graph[s].push_back(e);
    }
    memset(visited , false, sizeof visited);
    for (int i = 1 ; i <= n; i++){
        if(!visited[i]){
            toposort(i);
        }
    }
    for(int i = n-1;i >= 0 ; i--){
        cout << result[i] ;
        if (i != 0){
            cout << " ";
        }
    }

    return 0;
}

File: A/UVA/OrderingTasks.java
import java.util.ArrayList;
import java.util.Scanner;

class Main {
    static ArrayList<Integer> result =new ArrayList<>() ;
    public static void toposort(ArrayList<ArrayList<Integer>> graph,boolean []visited,int parent){
        visited[parent] = true;
        int sz = graph.get(parent).size();
        for(int i = 0;i<sz;i++){
            int child = graph.get(parent).get(i);
            if (!visited[child]){
                toposort(graph, visited, child);
            }
        }
        result.add(parent);
    }
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n =sc.nextInt(),m=sc.nextInt();
        sc.nextLine();
        ArrayList<ArrayList<Integer>> graph = new ArrayList<ArrayList<Integer>>(n+1);
        boolean []visited = new boolean[n+1];
        for(int i = 0 ;i <= n;i++){
            graph.add(new ArrayList<Integer>());
        } 
        int pre = sc.nextInt(), out = sc.nextInt();
        while(true){
            sc.nextLine();
            if(pre == 0 ) break;
            graph.get(pre).add(out);
            pre = sc.nextInt(); out = sc.nextInt();
        }
        for(int i = 1;i <=n ; i++){
            if(!visited[i]){
                toposort(graph,visited,i);
            }
        } 
        for(int i = n-1;i >= 0 ;i--)
        {
            System.out.print(result.get(i));
            if(i != 0 ){
                System.out.print(" ");
            }
        }
    }
}
